File: src/services/lark.ts
import config from 'config'
import axios from 'axios'
import streamifier from 'streamifier'
import * as lark from '@larksuiteoapi/node-sdk'
import type { NotifyService, SendResult, ChatItem } from './base.js'
import type { MessageParams, Channel } from '../schemas/notify.js'
import { logger } from '../lib/logger.js'
import { ServiceError } from '../lib/errors.js'

interface LarkConfig {
  appId: string
  appSecret: string
}

interface LarkCardMessage {
  config?: { wide_screen_mode?: boolean }
  header?: {
    title: { tag: string; content: string }
    template?: string
  }
  card_link?: { url: string }
  elements: Array<{
    tag: string
    [key: string]: unknown
  }>
}

export class LarkService implements NotifyService {
  readonly channel: Channel = 'lark'
  private client: lark.Client

  constructor() {
    const larkConfig = config.get<LarkConfig>('lark')
    this.client = new lark.Client({
      appType: lark.AppType.SelfBuild,
      domain: lark.Domain.Feishu,
      loggerLevel: lark.LoggerLevel.error,
      ...larkConfig
    })
  }

  async sendMessage(to: string, params: MessageParams): Promise<SendResult> {
    const message = this.buildCardMessage(params)
    return this.sendRawMessage(to, message)
  }

  async sendRawMessage(to: string, message: unknown): Promise<SendResult> {
    logger.info({ message, to }, 'Sending Lark message')

    const res = await this.client.im.message.create({
      params: { receive_id_type: 'chat_id' },
      data: {
        receive_id: to,
        msg_type: 'interactive',
        content: JSON.stringify(message)
      }
    }, lark.withTenantToken(''))

    if (res.code !== 0) {
      throw new ServiceError(`${res.code} - ${res.msg}`, 'lark')
    }

    return {
      messageId: res.data?.message_id,
      success: true
    }
  }

  async uploadImage(imageUrl: string): Promise<string> {
    const downloadResponse = await axios.get(imageUrl, { responseType: 'arraybuffer' })
    const imageBuffer = Buffer.from(downloadResponse.data)
    const imageReadableStream = streamifier.createReadStream(imageBuffer)

    const uploadResponse = await this.client.im.v1.image.create({
      data: {
        image_type: 'message',
        image: imageReadableStream as unknown as Buffer
      }
    })

    if (!uploadResponse?.image_key) {
      throw new ServiceError('Failed to upload image', 'lark')
    }

    return uploadResponse.image_key
  }

  async listChats(): Promise<ChatItem[]> {
    const res = await this.client.im.chat.list({
      params: { page_size: 100 }
    }, lark.withTenantToken(''))

    if (res.code !== 0) {
      throw new ServiceError(`${res.code} - ${res.msg}`, 'lark')
    }

    return (res.data?.items || []).map(item => ({
      chatId: item.chat_id!,
      name: item.name!,
      description: item.description
    }))
  }

  private buildCardMessage(params: MessageParams): LarkCardMessage {
    const message: LarkCardMessage = {
      config: { wide_screen_mode: true },
      elements: []
    }

    if (params.url) {
      message.card_link = { url: params.url }
    }

    if (params.title) {
      message.header = {
        title: { tag: 'plain_text', content: params.title },
        template: params.color || 'Blue'
      }
    }

    if (params.image) {
      message.elements.push({
        tag: 'img',
        img_key: params.image,
        alt: { tag: 'plain_text', content: params.title || 'image' }
      })
    }

    if (params.content) {
      message.elements.push({
        tag: 'markdown',
        content: params.content
      })
    }

    if (params.note) {
      if (params.content || params.url) {
        message.elements.push({ tag: 'hr' })
      }
      message.elements.push({
        tag: 'note',
        elements: [{ tag: 'plain_text', content: params.note }]
      })
    }

    return message
  }
}

File: internal/service/lark.go
package service

import (
	"bytes"
	"encoding/json"
	"fmt"
	"log/slog"
	"net/http"

	"notify/internal/config"
)

type LarkService struct {
	appID     string
	appSecret string
	client    *http.Client
}

func NewLarkService(cfg config.LarkConfig) *LarkService {
	return &LarkService{
		appID:     cfg.AppID,
		appSecret: cfg.AppSecret,
		client:    &http.Client{},
	}
}

func (s *LarkService) Channel() Channel {
	return ChannelLark
}

func (s *LarkService) BuildMessage(params MessageParams) any {
	return s.buildCardMessage(params)
}

func (s *LarkService) SendMessage(to string, params MessageParams) (*SendResult, error) {
	message := s.buildCardMessage(params)
	return s.SendRawMessage(to, message)
}

func (s *LarkService) SendRawMessage(to string, message any) (*SendResult, error) {
	slog.Info("Sending Lark message", "to", to, slog.Any("payload", message))

	token, err := s.getTenantAccessToken()
	if err != nil {
		return nil, fmt.Errorf("get tenant access token: %w", err)
	}

	content, _ := json.Marshal(message)
	reqBody := map[string]any{
		"receive_id": to,
		"msg_type":   "interactive",
		"content":    string(content),
	}

	body, _ := json.Marshal(reqBody)
	req, _ := http.NewRequest("POST", "https://open.feishu.cn/open-apis/im/v1/messages?receive_id_type=chat_id", bytes.NewReader(body))
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+token)

	resp, err := s.client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("send message: %w", err)
	}
	defer resp.Body.Close()

	var result struct {
		Code int    `json:"code"`
		Msg  string `json:"msg"`
		Data struct {
			MessageID string `json:"message_id"`
		} `json:"data"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("decode response: %w", err)
	}

	if result.Code != 0 {
		return nil, fmt.Errorf("lark error: %d - %s", result.Code, result.Msg)
	}

	return &SendResult{
		Success: true,
	}, nil
}


func (s *LarkService) ListChats() ([]ChatItem, error) {
	token, err := s.getTenantAccessToken()
	if err != nil {
		return nil, fmt.Errorf("get tenant access token: %w", err)
	}

	req, _ := http.NewRequest("GET", "https://open.feishu.cn/open-apis/im/v1/chats?page_size=100", nil)
	req.Header.Set("Authorization", "Bearer "+token)

	resp, err := s.client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("list chats: %w", err)
	}
	defer resp.Body.Close()

	var result struct {
		Code int    `json:"code"`
		Msg  string `json:"msg"`
		Data struct {
			Items []struct {
				ChatID      string `json:"chat_id"`
				Name        string `json:"name"`
				Description string `json:"description"`
			} `json:"items"`
		} `json:"data"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("decode response: %w", err)
	}

	if result.Code != 0 {
		return nil, fmt.Errorf("lark error: %d - %s", result.Code, result.Msg)
	}

	chats := make([]ChatItem, len(result.Data.Items))
	for i, item := range result.Data.Items {
		chats[i] = ChatItem{
			ChatID:      item.ChatID,
			Name:        item.Name,
			Description: item.Description,
		}
	}
	return chats, nil
}

func (s *LarkService) getTenantAccessToken() (string, error) {
	reqBody := map[string]string{
		"app_id":     s.appID,
		"app_secret": s.appSecret,
	}
	body, _ := json.Marshal(reqBody)

	resp, err := s.client.Post(
		"https://open.feishu.cn/open-apis/auth/v3/tenant_access_token/internal",
		"application/json",
		bytes.NewReader(body),
	)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	var result struct {
		Code              int    `json:"code"`
		Msg               string `json:"msg"`
		TenantAccessToken string `json:"tenant_access_token"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return "", err
	}

	if result.Code != 0 {
		return "", fmt.Errorf("get token failed: %d - %s", result.Code, result.Msg)
	}

	return result.TenantAccessToken, nil
}

func (s *LarkService) buildCardMessage(params MessageParams) map[string]any {
	message := map[string]any{
		"config":   map[string]any{"wide_screen_mode": true},
		"elements": []any{},
	}

	if params.URL != "" {
		message["card_link"] = map[string]any{"url": params.URL}
	}

	if params.Title != "" {
		color := params.Color
		if color == "" {
			color = ColorBlue
		}
		message["header"] = map[string]any{
			"title":    map[string]any{"tag": "plain_text", "content": params.Title},
			"template": string(color),
		}
	}

	elements := []any{}

	if params.Content != "" {
		elements = append(elements, map[string]any{
			"tag":     "markdown",
			"content": params.Content,
		})
	}

	if params.Note != "" {
		if params.Content != "" || params.URL != "" {
			elements = append(elements, map[string]any{"tag": "hr"})
		}
		elements = append(elements, map[string]any{
			"tag":      "note",
			"elements": []any{map[string]any{"tag": "plain_text", "content": params.Note}},
		})
	}

	message["elements"] = elements
	return message
}
